python/lsst/daf/butler/registry/datasets/byDimensions/_storage.py

```diff
@@ -77,20 +77,36 @@ def __init__(
         collections: CollectionManager,
         static: StaticDatasetTablesTuple,
         summaries: CollectionSummaryManager,
-        tags: sqlalchemy.schema.Table,
+        tags_table_factory: Callable[[], sqlalchemy.schema.Table],
         use_astropy_ingest_date: bool,
-        calibs: sqlalchemy.schema.Table | None,
+        calibs_table_factory: Callable[[], sqlalchemy.schema.Table] | None,
     ):
         super().__init__(datasetType=datasetType)
         self._dataset_type_id = dataset_type_id
         self._db = db
         self._collections = collections
         self._static = static
         self._summaries = summaries
-        self._tags = tags
-        self._calibs = calibs
+        self._tags_table_factory = tags_table_factory
+        self._calibs_table_factory = calibs_table_factory
         self._runKeyColumn = collections.getRunForeignKeyName()
         self._use_astropy = use_astropy_ingest_date
+        self._tags_table: sqlalchemy.schema.Table | None = None
+        self._calibs_table: sqlalchemy.schema.Table | None = None
+
+    @property
+    def _tags(self) -> sqlalchemy.schema.Table:
+        if self._tags_table is None:
+            self._tags_table = self._tags_table_factory()
+        return self._tags_table
+
+    @property
+    def _calibs(self) -> sqlalchemy.schema.Table | None:
+        if self._calibs_table is None:
+            if self._calibs_table_factory is None:
+                return None
+            self._calibs_table = self._calibs_table_factory()
+        return self._calibs_table
```

Comment:
This is already an improvement as-is, but if we moved this pattern of only looking up the table on first use up a level, so that we only make a `DatasetRecordStorage` instance at all on first use, it might help us get the bad kind of caching out of that manager so we can use it on the server.  Happy to defer that work to another ticket; just figured it was worth mentioning here because it might make a lot of this change go away.

Reply:
I think there is a reason to delay table reflection even if we create DatasetRecordStorage instances on-demand. `resolve_wildcard` needs all dataset types when patters are used, so we'd probably need to make all storages at that point, but we still don't need to know their tables' schemas. Alternative is probably to just query dataset types on each `resolve_wildcard` without making storage instances?

Reply:
We could also load the list of dataset type names to do the pattern-matching without necessarily loading the storages at the same time, in the same way we talked about decoupling loading the children-of-chains names from loading the records of those children.

Reply:
OK, I added another commit [8d3458f](https://github.com/lsst/daf_butler/pull/904/commits/8d3458fade951351207fd051182261fe6e3f4439) which completely eliminates caching from datasets manager. It makes summaries manager a bit messier and more dependent on datasets manager (and dataset_type table). Could you check it and tell if you like this variant better?

Reply:
It's certainly more viable for use in the server, so I think we'll have to do something like this (or something more drastic in terms of refactoring managers) at some point.  I don't think the manager system is really serving us well at this point, so I'm not too bothered by the collection and dataset managers getting more dependent on each other - it just feels like yet another indication of something we already knew, or at least suspected.

I am a little worried that stripping the dataset type caching out of the managers without putting it somewhere else in `DirectButler` could be a noticeable regression for certain use cases, but it's probably okay for now since you should be able to claim a much more obvious speedup in `DirectButler` initialization times from this ticket.

Reply:
I think I need to re-introduce some caching, at Buter level. I checked how many queries graph builder generates for ci_hsc_gen3 and there is a large increase compared to fully-cached case (timing is not very different, but this was with local postgres).